Gateway/Response/PaymentDetailsHandler.php

```diff
@@ -31,16 +66,17 @@ public function handle(array $handlingSubject, array $response)
         $payment->setAdditionalInformation('charge_id', $response['charge']->id);
         $payment->setAdditionalInformation('charge_authorize_uri', $response['charge']->authorize_uri);
         $payment->setAdditionalInformation('payment_type', $response['charge']->source['type']);
-        
-        if ($response['charge']->source['type'] === 'bill_payment_tesco_lotus') {
-            $barcode = $this->downloadPaymentFile($response['charge']->source['references']['barcode']);
-            $payment->setAdditionalInformation('barcode', $barcode);
+        $paymentType = $response['charge']->source['type'];
+        if ($paymentType === 'bill_payment_tesco_lotus') {
+            $this->setPaymentFileData($response['charge']->source['references']['barcode']);
+            $payment->setAdditionalInformation('barcode', $this->imageData);
+            return;
         }
 
-        if ($response['charge']->source['type'] === 'paynow') {
-            $qrCodeImage = $this->downloadPaymentFile($response['charge']->source['scannable_code']['image']['download_uri']);
-            $payment->setAdditionalInformation('qr_code_encoded', base64_encode($qrCodeImage));
+        if ($this->_helper->isPayableByImageCode($paymentType)) {
+            $this->setPaymentFileData($response['charge']->source['scannable_code']['image']['download_uri']);
+            $payment->setAdditionalInformation('qr_code_encoded', base64_encode($this->imageData));
+            $payment->setAdditionalInformation('qr_data_type', $this->imageType);
```

Comment:
(optional, just idea) this could be organise more by separating these part of code out from `PaymentDetailsHandler` to, maybe `BarcodeHandler` class and inject back via `di.xml` for each of payment method (applying SOLID here without having to check `if ($this->_helper->isPayableByImageCode($paymentType)) { }`).